pkg/sql/catalog/singleversion/singleversion.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+// Package singleversion provides abstractions to Lease descriptors in a
+// way which prevents new versions from being introduced.
+//
+// This layer enables changefeeds to operate at low-latency.
+package singleversion
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+)
+
+// Lease represents a lease which prevents a descriptor from being modified.
+type Lease interface {
+	Start() hlc.Timestamp
+	Expiration() hlc.Timestamp
+	Release(ctx context.Context, maxUsedTimestamp hlc.Timestamp)
+	ID() descpb.ID
+}
+
+// Acquirer is used to acquire a lease.
+type Acquirer interface {
+
+	// Acquire acquires a new singleversion lease.
+	Acquire(ctx context.Context, id descpb.ID) (Lease, error)
+
+	// WatchForNotifications watches for all notifications to singleversion
+	// leases destined for the specified IDs.
+	WatchForNotifications(
```

Comment:
It's a bit .. hand wavy... Just thinking out loud if you will.
The WatchForNotification implementation starts this dance with underlying svstorage to start Rangefeed over
"/action/id/session" span, waiting for events to come in.  The purpose of this method is to be called by somebody else (changefeed system, in particular), to establish a rangefeed that reacts to certain events, and effectively invalidates the lease that the changefeed has..

There is quite a bit of complexity here.  Also, we are establishing a separate rangefeed for each changefeed running on the system... However, at this point, changefeed is the only system at play.  And even if it weren't, the single version leases are not intended to be huge -- and are certainly bound by number of nodes * number of descriptors -- i.e. probably not that many.

So.. I guess I wonder if we should instead have storage itself establish a rangefeed for the entirety of the system table?
Acquire would return Lease, and the returned lease will add itself to the list of leases that ought to be notified when events happen.  Notification could mean marking the lease invalid, etc...

I think from the callers perspective, this might be nice because all you need to do is to Acquire the lease, and make sure you Release() it.  You can even request that Lease to refresh itself if it became invalidated...

As I said.. hand wavy, but want to bring this to your attention.

Reply:
The thing that's delicate here is making sure that you see all notifications. It's really hard for me to see how to share an underlying rangefeed without some buffering which I don't want to have to think about. Like, imagine that an event gets received at ts 2 and then at ts1 on the local node a rangefeed starts but misses the ts2 event. It will have no way to know that it missed the event. 

I will say that at first this created one rangefeed per descriptor, but that felt worse to me. 

At a deeper level, it'd be cool if we had some joining protocol whereby there was one stream between each server and each remote range, but that's some serious optimization territory.